nengo_dl/simulator.py

```diff
@@ -117,15 +117,16 @@ def __init__(self, network, dt=0.001, seed=None, model=None,
 
         # TODO: multi-GPU support
 
-        if device is None:
-            # check GPU support
-            installed_dists = [d.project_name for d in
-                               pkg_resources.working_set]
-            if ("tensorflow-gpu" not in installed_dists and
-                    "tf-nightly-gpu" not in installed_dists):
-                warnings.warn(
-                    "No GPU support detected. It is recommended that you "
-                    "install tensorflow-gpu (`pip install tensorflow-gpu`).")
+        installed_dists = [d.project_name for d in pkg_resources.working_set]
+        if device is None and ("tensorflow-gpu" not in installed_dists and
+                               "tf-nightly-gpu" not in installed_dists):
+            warnings.warn(
+                "No GPU support detected. It is recommended that you "
+                "install tensorflow-gpu (`pip install tensorflow-gpu`).")
+            logger.info("Running on CPU")
+        else:
+            logger.info("Running on %s", "CPU/GPU" if device is None else (
```

Comment:
I would maybe put "Unknown device" or something if `device is None`. "CPU/GPU" makes it sound like it's using both the CPU and GPU.

Reply:
That is what happens (if you don't explicitly specify a device TensorFlow will assign ops to devices according to internal rules, which will likely involve a mix of CPU and GPU execution).

Reply:
Ahh, I see. I just want to be sure that the GPU is definitely being used in that case. Is it possible for `tensorflow-gpu` to be installed, but e.g. CUDA not be set up right so the GPU doesn't actually get used?

Reply:
If you have `tensorflow-gpu` installed then it will try to load the CUDA libraries on import, and you'll get an error if it can't find them (just double checked that that is true).  Same if you have an incompatible version of CUDA installed.  So I'm pretty sure you can be confident that if you have `tensorflow-gpu` installed and you aren't seeing any errors, it's using the GPU.  

It's possible that there could be some obscure way to break it past that?  But that'd be a situation where TensorFlow believes it is running on the GPU, but it isn't actually running on the GPU.  TensorFlow doesn't silently defer to the CPU in case of errors, it just throws the error it seems.

As a sidenote, if you'd like to see which ops are assigned to which device, you can use

```
nengo_dl.configure_settings(session_config={"log_device_placement": True})
```

Also note that if anyone ever wants these log messages back they can be re-enabled via
```
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"
```

Reply:
A slight amendment to that.  I think it might be possible to have an error where CUDA is installed but it isn't detecting your GPU.  In this case TensorFlow will assign all the ops to the only available device (CPU).  However, if you explicitly assign ops to a GPU device, you will get an error if none is detected.  So I should revise the statement above to:

> You can be confident that if you aren't seeing any errors when running with `device="/gpu:0"`, it's using the GPU.

And then you can also be sure that if you can run models with `device="/gpu:0"`, `device=None` can also run on the GPU (although it may not, depending on TensorFlow's internal placement logic).